### PR TITLE
HandshakeStoppedState, ReceiveStoppedState, ReceiveFailedState, ReceivingState and ReceiveReconnectState transitions

### PR DESCRIPTION
**Pull request by budgetpreneur**

### CODE REVIEW

File: src/Api/PubnubApi/EventEngine/Subscribe/Events/SubscriptionEvents.cs [hunk 1]
```diff
@@ -40,11 +40,15 @@ public class HandshakeReconnectGiveUpEvent : Core.IEvent {
 	}
 
 	public class ReceiveSuccessEvent : Core.IEvent {
-		public List<PNMessageResult<object>> messages;
-		public SubscriptionCursor cursor;
+		public IEnumerable<string> Channels;
```
Comment (budgetpreneur) [inline src/Api/PubnubApi/EventEngine/Subscribe/Events/SubscriptionEvents.cs:43]: per CE decisions document, `RECEIVE_SUCCESS` event do not have `Channels` and `ChannelGroups`. If we don't have, how can ReceiveState get info of `Channels` and `ChannelGroups`, which expects the data from `ReceiveSuccessEvent`

Comment (budgetpreneur) [reply src/Api/PubnubApi/EventEngine/Subscribe/Events/SubscriptionEvents.cs:43]: Probably Channels and ChannelGroups are declared in global context.

Comment (MikeDobrzan) [reply src/Api/PubnubApi/EventEngine/Subscribe/Events/SubscriptionEvents.cs:43]: Agreed, the document is missing that.

File: src/Api/PubnubApi/EventEngine/Subscribe/Invocations/SubscriptionInvocations.cs [hunk 2]
```diff
@@ -37,7 +37,12 @@ internal class HandshakeReconnectInvocation: Core.IEffectInvocation
 
 	internal class CancelHandshakeReconnectInvocation: HandshakeReconnectInvocation, Core.IEffectCancelInvocation { }
 	
-	internal class ReceiveReconnectInvocation: Core.IEffectInvocation { }
+	internal class ReceiveReconnectInvocation: Core.IEffectInvocation 
+	{ 
+		public IEnumerable<string> Channels;
+		public IEnumerable<string> ChannelGroups;
+	}
```
Comment (MikeDobrzan) [inline src/Api/PubnubApi/EventEngine/Subscribe/Invocations/SubscriptionInvocations.cs:44]: I think it needs to pass the Cursor

File: src/Api/PubnubApi/EventEngine/Subscribe/States/ReceivingState.cs [hunk 3]
```diff
@@ -12,7 +12,70 @@ internal class ReceivingState : Core.IState {
...
+						new[] {
+							new ReceiveMessagesInvocation() {
+								Channels = subscriptionChanged.Channels,
+								ChannelGroups = subscriptionChanged.ChannelGroups,
+							},
+						}
+					);
+				case Events.SubscriptionRestoredEvent subscriptionRestored:
+					return new Tuple<IState, IEnumerable<IEffectInvocation>>(
+						new HandshakeFailedState() {
```
Comment (MikeDobrzan) [inline src/Api/PubnubApi/EventEngine/Subscribe/States/ReceivingState.cs:44]: It should do a self-transistion

File: src/Api/PubnubApi/EventEngine/Subscribe/Invocations/SubscriptionInvocations.cs (hunk 2 above)
Comment (budgetpreneur) [reply src/Api/PubnubApi/EventEngine/Subscribe/Invocations/SubscriptionInvocations.cs:46]: Added cursor

File: src/Api/PubnubApi/EventEngine/Subscribe/States/ReceivingState.cs (hunk 3 above)
Comment (budgetpreneur) [reply src/Api/PubnubApi/EventEngine/Subscribe/States/ReceivingState.cs:49]: Added cursor

**Review by MikeDobrzan — CHANGES_REQUESTED** [review]

File: src/Api/PubnubApi/EventEngine/Subscribe/States/ReceivingState.cs [hunk 4]
```diff
@@ -8,11 +8,83 @@ internal class ReceivingState : Core.IState {
...
+						},
+						new[] {
+							new ReceiveMessagesInvocation() {
+								Channels = subscriptionChanged.Channels,
+								ChannelGroups = subscriptionChanged.ChannelGroups,
+								Cursor = subscriptionChanged.Cursor
+							},
+						}
+					);
+				case Events.SubscriptionRestoredEvent subscriptionRestored:
```
[the remark this answers is not in the document]
Comment (budgetpreneur) [reply src/Api/PubnubApi/EventEngine/Subscribe/States/ReceivingState.cs:47]: fixed typo to ReceivingState

**Review by MikeDobrzan — APPROVED** [review]

**Pull request merged**